autoprecompiles/src/optimizer.rs

```diff
@@ -112,60 +105,79 @@ fn solver_based_optimization<T: FieldElement>(
     result.simplified_constraint_system
 }
 
-fn remove_trivial_bus_interactions<T: FieldElement>(
+/// Removes any columns that are not connected to *stateful* but interactions (e.g. memory),
+/// because those are the only way to interact with the rest of the system.
+/// We assume that all input constraints are satisfiable. Because they are not connected to
+/// the state of the system, the prover can always satisfy them, so removing them is safe.
+/// Note that if there was an unsatisfiable constraints, it might also be removed, which would
+/// change the statement being proven.
+fn remove_disconnected_columns<T: FieldElement>(
     constraint_system: ConstraintSystem<T, Variable>,
-    bus_interaction_handler: impl ConcreteBusInteractionHandler<T> + 'static,
+    bus_interaction_handler: impl IsBusStateful<T>,
 ) -> ConstraintSystem<T, Variable> {
-    let ConstraintSystem {
-        algebraic_constraints,
-        bus_interactions,
-    } = constraint_system;
+    // Initialize vars_to_keep with any variables that appear in stateful bus interactions.
+    let mut variables_to_keep = constraint_system
+        .bus_interactions
+        .iter()
+        .filter(|bus_interaction| {
+            let bus_id = bus_interaction.bus_id.try_to_number().unwrap();
+            bus_interaction_handler.is_stateful(bus_id)
+        })
+        .flat_map(|bus_interaction| bus_interaction.referenced_variables())
+        .cloned()
+        .collect::<HashSet<_>>();
+
+    // Any variable that is connected to a variable in vars_to_keep must also be kept.
+    loop {
+        let size_before = variables_to_keep.len();
+        for constraint in &constraint_system.algebraic_constraints {
+            if constraint
+                .referenced_variables()
+                .any(|var| variables_to_keep.contains(var))
+            {
+                variables_to_keep.extend(constraint.referenced_variables().cloned());
+            }
+        }
+
+        for constraint in &constraint_system.bus_interactions {
```

Comment:
I think `SymbolicMachine` has an iterator function that traverses all polynomial constraints and bus interactions. Maybe the same thing can be done for `ConstraintSystem`? Then these two blocks could be collapsed into a single one, since they're equal but run over different objects.